### ISSUE TITLE
Wrong type argument when union of collections is passed to generic function 

### ISSUE THREAD

**Issue by laikoni** [microsoft/pyright#4121]
**Describe the bug**
Assume the variable `x` has type `list[int] | list[str]`, so it is either a list containing only ints or a list containing only strings.

Then `set(x)` is either a set containing only ints, or a set containing only strings, so I would expect it to have the type `set[int] | set[str]`. However, Pylance infers the type `set[int | str]`, which describes a set that may contain _both_ ints and strings, so clearly a different type. (Indeed, Pylance then agrees that these two types are not compatible.)

**Example**

```python
def my_set(x: list[int] | list[str]) -> set[int] | set[str]:
    return set(x)
```
Now Pylance reports at `set(x)`:

```python
Expression of type "set[int | str]" cannot be assigned to return type "set[int] | set[str]"
  Type "set[int | str]" cannot be assigned to type "set[int] | set[str]"
    "set[int | str]" is incompatible with "set[int]"
      TypeVar "_T@set" is invariant
        Type "int | str" cannot be assigned to type "int"
          "str" is incompatible with "int"
    "set[int | str]" is incompatible with "set[str]"
      TypeVar "_T@set" is invariant
        Type "int | str" cannot be assigned to type "str"
  ...Pylance(reportGeneralTypeIssues)
```

**VS Code extension or command-line**
Pylance v2022.11.10
Python 3.11.0

**Additional context**
mypy also reports a type error for the above example, however it inferes the type `set[object]` for `set(x)` (see this [issue](https://github.com/python/mypy/issues/13960)). It is mentioned there that this behaviour is likely caused by mypy using joins instead of unions, something that pyright apparently does not do, so there appears to be a different issue here.

**Comment by erictraut** [comment]
I can understand why you think that the current behavior is undesirable, but it is correct from a type perspective.

This case doesn't involve type inference. It involves a different type checking concept called "constraint solving". The input parameter to the `set` constructor has the type `Iterable[_T]`. The type `list[int] | list[str]` is type compatible with `Iterable[_T]`. When pyright solves constraints (i.e. TypeVars), it uses a union operation. By contrast, mypy uses a join operation. Both are permissible, but we've found that unions produce better results almost all of the time. Mypy is considering changing from join to union as well. In this example, `_T` is solved as either `int | str` or `object` depending on whether a union or join is used. Both solutions are correct.

I don't see a good way to work around this limitation unfortunately.
